### PR TITLE
Feat/sushiswap

### PR DESCRIPTION
**Pull request by vic-en**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
This pr adds SushiSwap support on BSC chain.

**Tests performed by the developer**:
- Ensured all test pass locally.
- Did basic endpoint testing and a bit of end-to-end connection from the client.

**Tips for QA testing**:
- Ensure all new curl tests work
- End-to-end testing with AMM arb strategy and ensure there are no issues.

### CODE REVIEW

**Comment by JeremyKono** [comment]
@vic-en please resolve conflcits. Thanks

**Comment by rapcmia** [comment]
PR update:
- Successfully setup docker gateway build for endpoint tests
- Endpoint tests on Curl
  - Setup BSC env variables 
  - port: ok 
  - wallet/add: ok 
  - wallet/remove: ok
  - network/balances: ok 
  - evm/nonce: ok 
  - evm/approve: ok 
  - amm/price
  - amm/trade
  
However not able to test hummingbot client since its returning the error of `Unhandled error in background task: local variable 'wallet_account_id'`  that was introduce on the our development branch but the PR to fix it has been deployed today. We assume we need to update the PR to latest development branch

Screenshot 2022-11-17 153115

```
2022-11-17 15:30:44,149 - 30619 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: local variable 'wallet_account_id' referenced before assignment
Traceback (most recent call last):
  File "/home/rapcmia/github/5984/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/home/rapcmia/github/5984/hummingbot/client/command/gateway_command.py", line 431, in _gateway_connect
    chain, network, wallet_private_key, id=wallet_account_id
UnboundLocalError: local variable 'wallet_account_id' referenced before assignment
```

**Comment by nikspz** [comment]
@vic-en Could you please resolve branch conflicts? Currently blocked to test Client with it

**Comment by vic-en** [comment]
@nikspz all done

**Comment by rapcmia** [comment]
Hi @vic-en good day,
Would you please fix the branch conflicts when available? Thank you

File: gateway/test/services/wallet/wallet.controllers.test.ts [hunk 1]
```diff
@@ -20,13 +20,15 @@ import { ConfigManagerCertPassphrase } from '../../../src/services/config-manage
 import { BinanceSmartChain } from '../../../src/chains/binance-smart-chain/binance-smart-chain';
 import { Cronos } from '../../../src/chains/cronos/cronos';
 import { Near } from '../../../src/chains/near/near';
+import { BinanceSmartChain } from '../../../src/chains/binance-smart-chain/binance-smart-chain';
```
Comment (vic-en) [inline gateway/test/services/wallet/wallet.controllers.test.ts:23]:
```suggestion
```

File: gateway/src/services/wallet/wallet.controllers.ts [hunk 2]
```diff
@@ -2,6 +2,7 @@ import fse from 'fs-extra';
 import { Avalanche } from '../../chains/avalanche/avalanche';
 import { BinanceSmartChain } from '../../chains/binance-smart-chain/binance-smart-chain';
 import { Cronos } from '../../chains/cronos/cronos';
+import { BinanceSmartChain } from '../../chains/binance-smart-chain/binance-smart-chain';
```
Comment (vic-en) [inline gateway/src/services/wallet/wallet.controllers.ts:5]:
```suggestion
```

File: gateway/src/services/wallet/wallet.controllers.ts [hunk 3]
```diff
@@ -59,6 +59,8 @@ export async function addWallet(
     connection = Solana.getInstance(req.network);
   } else if (req.chain === 'polygon') {
     connection = Polygon.getInstance(req.network);
+  } else if (req.chain === 'binance-smart-chain') {
+    connection = BinanceSmartChain.getInstance(req.network);
```
Comment (vic-en) [inline gateway/src/services/wallet/wallet.controllers.ts:63]:
```suggestion
```

File: gateway/test/services/wallet/wallet.controllers.test.ts [hunk 4]
```diff
@@ -27,6 +27,7 @@ let eth: Ethereum;
 let harmony: Harmony;
 let bsc: BinanceSmartChain;
 let near: Near;
+let bsc: BinanceSmartChain;
```
Comment (vic-en) [inline gateway/test/services/wallet/wallet.controllers.test.ts:30]:
```suggestion
```

File: gateway/test/services/wallet/wallet.controllers.test.ts [hunk 5]
```diff
@@ -50,6 +51,7 @@ afterAll(async () => {
   await bsc.close();
   await cronos.close();
   await near.close();
+  await bsc.close();
```
Comment (vic-en) [inline gateway/test/services/wallet/wallet.controllers.test.ts:54]:
```suggestion
```

File: gateway/test/services/wallet/wallet.controllers.test.ts [hunk 6]
```diff
@@ -212,6 +214,32 @@ describe('addWallet and getWallets', () => {
...
+
+    const wallets = await getWallets();
+
+    const addresses: string[][] = wallets
+      .filter((wallet) => wallet.chain === 'binance-smart-chain')
+      .map((wallet) => wallet.walletAddresses);
+
+    expect(addresses[0]).toContain(oneAddress);
+  });
+
```
Comment (vic-en) [inline gateway/test/services/wallet/wallet.controllers.test.ts:242]:
```suggestion
```

**Comment by rapcmia** [comment]
PR update: 
- Successfully setup source + source gateway build
- Successfully connected wallet for sushiswap
- Tests other wallets uniswap an pancakeswap, all ok 
- Setup endpoint test using curl commands
   - Setup env variables
   - Port, wallet/add, wallet/remove, network/balances, evm/nonce, evm/approve, amm/price and amm/trade all ok 
   - Compare results from endpoint and [BSCscan](https://bscscan.com/), matched

Pending:
- Client tests on source + source gateway build
- Client tests on docker + docker gateway build

**Review by rapcmia — APPROVED** [review]
Test summary:
- Setup successfully on both source and docker gateway build
- Successfully connected Pancakeswap wallet and balance displays correct
- Test other chain/network and setup wallet successfully
- Connector-tokens feature also is displaying the tokens relevant to the connector
- Setup simple AMM-arbitrage with Sushiswap
- Runtime of more than 45mins~ no error found
- Rate probability is working as expected as no observed unrealistic rates (e.g 303%, -94% etc)
- Order execution:
   - Multiple order event triggered, all recorded and txhash show successfully swap
   - Set min_profitability for interval arbitrage opporunity. Trades successfully recorded
   - When balance not enough, the client return below required order amount
   - Trades matched from txhash and client history/csv files
- Tested on bot source and docker client build both running on AWS and local instances
- Successfully tested endpoints via curl commands

**Pull request merged**